juju/juju
# feature: move-to-space: Apiserver part

### Checklist


----

## Description of change

- adds the apiserver part to move a list of CIDRs to a space from their old space.
- the cmd part will be implemented in another patch. The command will look like this:
```
juju move-to-space <spacename> <sliceOfCIDRs>
```

- **Additionally, this patch does not include the check and movement in case of endpoint bindings which are used and may or may not be moved as well. This work will be added in another patch.**

#### Possible cases which the code currently handles:
- [ ] Wrong CIDR (2)
- [ ] CIDR/Subnet does not exist (3)
- [ ] Space does not exist (4)
- [ ] CIDR is already on space (5)
- [ ] (1) but with force option (6)
- [ ] Success case (7)
- [ ] Moving a CIDR of a used machine to another space, which has a negative constraint. [8]
- [ ] Machine has an address on that subnet, but no constraint is blocking. [9]

## QA steps
There are no specific QA steps doable for now, as the CMD part is missing.
If someone wants to try the changes locally one can apply checkout the branch: https://github.com/nammn/juju/tree/updatespace-with-cmd

With this, there should be a basic cmd implementation to the move-to-space implementation 

[2] wrong CIDR
```
❯ juju move-to-space dsadas 132.31.0.32130/28
ERROR invalid arguments specified: "132.31.0.32130/28" is not a valid CIDR
```
[3] subnet not found
```
~
❯ juju move-to-space dsadas 132.31.0.0/28
ERROR cannot update space "dsadas": subnet "132.31.0.0/28" not found
```

[4] space not found
```
~
❯ juju move-to-space dsadas 172.31.0.0/28
ERROR cannot update space "dsadas": space "dsadas" not found
```

[5] already in space
```
~
❯ juju move-to-space db 172.31.0.0/28 
ERROR cannot update space "db": supplied CIDR "172.31.0.0/28" is already in space "db"
```


[7] basic success case
```
❯ juju deploy mediawiki 
~
❯ juju move-to-space db 172.31.0.0/28
Subnet "172.31.0.0/28" moved from "alpha" to "db"
```
<- note that 172.31.0.0/28 is not used on any machine 

[8] is more complex, I will add this as a step QA step here:
```
❯ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/28 
                 172.31.1.0/28 
                 172.31.16.0/20
                 172.31.2.0/28 
                 172.31.3.0/24 
db     1  
❯ juju deploy mediawiki --constraints spaces=alpha,^db
❯ juju move-to-space db 172.31.16.0/20
ERROR cannot update space "db": cannot move CIDR "172.31.16.0/20" from space "alpha" to space: "^db", as this would violate the current application constraint: "^db" on application "mediawiki".
```

[9] subnet in use, but no constraint

`juju deploy mysql`
... wait until machine, which mysql get deployed to, is online. In this case, it is machine "1"
`juju show-machine 1 `
--> look for the ip address matching a CIDR from a space. In my example, it is 172.31.26.60
```
❯ juju move-to-space db 172.31.16.0/20
Subnet "172.31.16.0/20" moved from "alpha" to "db"
```



## Spec
https://docs.google.com/document/d/1tE5GMF9Uw8W7QQoybgA_vs_RqCHQdbKXmQT_Kaaj6Pk/edit#heading=h.fro54rq52v4
